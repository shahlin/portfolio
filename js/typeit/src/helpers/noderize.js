import convertHTMLEntities from "./convertHTMLEntities";

const PLACEHOLDER_PATTERN = "{%}";

/**
 * Replace nodes with string placeholders.
 *
 * @param string string
 */
export function placeholderize(string) {
  let parser = new DOMParser();
  let doc = parser.parseFromString(string, "text/html");
  let nodes = [].slice.call(doc.body.querySelectorAll("*"));

  //-- Remove empty tags.
  nodes = nodes.filter(item => {
    //-- This thing is empty.
    if (item.outerHTML.indexOf("></") > -1) {
      string = string.replace(item.outerHTML, "");
      return false;
    }

    return true;
  });

  //-- Replace node strings with placeholders. I'm not concerned with WHICH -- just want them all converted.
  nodes.forEach(item => {
    //-- Accounts for standard AND self closing tags (note the optional content & closing tag).
    string = string.replace(
      new RegExp(`<${item.tagName}(.*?)\/?>((.*?)<\/${item.tagName}>)?`, "i"),
      PLACEHOLDER_PATTERN
    );
  });

  return {
    string,
    nodes
  };
}

/**
 * Remove remants "%}" of placeholders that remain
 * after noderizing an array.
 *
 * @param array items
 */
export function removePlaceholderRemnants(items) {
  let isPruning = true;

  while (isPruning) {
    let hasLastCharacters = items.some((item, index) => {
      let isLastNodeObject =
        typeof item === "object" &&
        (item.isLastCharacter || item.content === null);

      if (
        isLastNodeObject &&
        items.slice(index + 1, index + 3).join("") === "%}"
      ) {
        items.splice(index + 1, 2);
        return true;
      }

      return false;
    });

    if (!hasLastCharacters) {
      isPruning = false;
    }
  }

  return items;
}

export default function(rawString) {
  let { string, nodes } = placeholderize(rawString);

  let stringArray = convertHTMLEntities(string).split("");
  let nodifiedArray = [];

  stringArray.forEach((item, index) => {
    //-- Just a regular character.
    if (stringArray.slice(index, index + 3).join("") !== PLACEHOLDER_PATTERN) {
      nodifiedArray.push(item);
      return;
    }

    //-- Replace placeholder w/ node objects.
    let firstCharacterIndex = index;
    let node = nodes.shift();
    let nodeContents = convertHTMLEntities(node.innerHTML).split("");
    let nodeAttributes = [].slice.call(node.attributes).map(att => {
      return {
        name: att.name,
        value: att.nodeValue
      };
    });

    if (nodeContents.length) {
      nodeContents.forEach((character, i) => {
        nodifiedArray.push({
          tag: node.tagName,
          attributes: nodeAttributes,
          content: character,
          isFirstCharacter: firstCharacterIndex === index,
          isLastCharacter: i + 1 === nodeContents.length
        });

        firstCharacterIndex++;
      });
    } else {
      nodifiedArray.push({
        tag: node.tagName,
        attributes: nodeAttributes,
        content: null
      });
    }
  });

  nodifiedArray = removePlaceholderRemnants(nodifiedArray);

  return nodifiedArray;
}
