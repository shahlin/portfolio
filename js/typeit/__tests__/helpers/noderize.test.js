import noderize, {
  placeholderize,
  removePlaceholderRemnants
} from "../../src/helpers/noderize";

test("Parses normal string correctly.", () => {
  let result = noderize("Hello, this is my string.");

  expect(result).toMatchSnapshot();
});

test("Parses single HTML tag.", () => {
  let result = noderize("Hello, this is some <strong>bold</strong> text.");

  expect(result).toMatchSnapshot();
});

test("Parses multiple HTML tags.", () => {
  let result = noderize(
    "Hello, this is some <strong>bold</strong> text, and some <i>italicized</i> text."
  );

  expect(result).toMatchSnapshot();
});

test("Parses HTML tag at beginning of string.", () => {
  let result = noderize(
    "<strong>Hello!</strong> This is some text with HTML at the beginning."
  );

  expect(result).toMatchSnapshot();
});

test("Parses HTML tag at end of string.", () => {
  let result = noderize(
    "This is some text with HTML at the <em>beginning.</em>"
  );

  expect(result).toMatchSnapshot();
});

test("Parses HTML tag with attributes.", () => {
  let result = noderize(
    'This string has an <strong class="strong-class" id="strong-id" data-whatever="data-att">element</strong> with attributes.'
  );

  expect(result).toMatchSnapshot();
});

describe("placeholderize()", () => {
  test("Sets placeholder for single regular HTML tag.", () => {
    let result = placeholderize(
      "This is a string with some <strong>bold text</strong>."
    );

    expect(result).toMatchSnapshot();
  });

  test("Sets placeholder for a self-closing tag.", () => {
    let result = placeholderize("This is a string with a <br/> break in it.");

    expect(result).toMatchSnapshot();
  });

  test("Sets placeholder for a self-closing tag with no closing slash.", () => {
    let result = placeholderize("This is a string with a <br> break in it.");

    expect(result).toMatchSnapshot();
  });

  test("Sets placeholder for strings with both standard tags and self-closing.", () => {
    let result = placeholderize(
      "This is a string with a <br> break in it, as well as some <em>italized</em> text."
    );

    expect(result).toMatchSnapshot();
  });

  test("Removes empty tags.", () => {
    let result = placeholderize("<span></span>");

    expect(result).toMatchSnapshot();
  });

  test("Removes empty tags with text surrounding them.", () => {
    let result = placeholderize("text! <span></span> text!");

    expect(result).toMatchSnapshot();
  });

  test("Removes empty tags that live with filled tags.", () => {
    let result = placeholderize("<span></span> <em>italics!</em>");

    expect(result).toMatchSnapshot();
  });
});

describe("removePlaceholderRemnants()", () => {
  test("It removes placeholder remnants following the end of a tag.", () => {
    let payload = [
      "x",
      {
        isLastCharacter: false
      },
      {
        isLastCharacter: false
      },
      {
        isLastCharacter: true
      },
      "%",
      "}"
    ];

    expect(removePlaceholderRemnants(payload)).toMatchSnapshot();
  });

  test("It does not touch array if last character is not followed by remnants.", () => {
    let payload = [
      "x",
      {
        isLastCharacter: false
      },
      {
        isLastCharacter: false
      },
      {
        isLastCharacter: true
      },
      "%",
      "%"
    ];

    expect(removePlaceholderRemnants(payload)).toMatchSnapshot();
  });

  test("It properly handles arrays whose node objects are at the very beginning.", () => {
    let payload = [
      {
        isLastCharacter: false
      },
      {
        isLastCharacter: false
      },
      {
        isLastCharacter: true
      },
      "%",
      "%"
    ];

    expect(removePlaceholderRemnants(payload)).toMatchSnapshot();
  });

  test("It properly handles arrays completely filled with node objects.", () => {
    let payload = [
      {
        isLastCharacter: false
      },
      {
        isLastCharacter: false
      },
      {
        isLastCharacter: true
      }
    ];

    expect(removePlaceholderRemnants(payload)).toMatchSnapshot();
  });

  test("It does not modify arrays with no node objects.", () => {
    let payload = ["a", "b", "c"];

    expect(removePlaceholderRemnants(payload)).toEqual(["a", "b", "c"]);
  });

  test("It does not modify arrays with no node objects.", () => {
    let payload = [
      "a",
      {
        tag: "BR",
        attributes: [],
        content: null
      },
      "%",
      "}"
    ];

    expect(removePlaceholderRemnants(payload)).toEqual([
      "a",
      { tag: "BR", attributes: [], content: null }
    ]);
  });
});
