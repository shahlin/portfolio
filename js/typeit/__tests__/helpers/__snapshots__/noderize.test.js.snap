// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parses HTML tag at beginning of string. 1`] = `
Array [
  Object {
    "attributes": Array [],
    "content": "H",
    "isFirstCharacter": true,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "e",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "l",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "l",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "o",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "!",
    "isFirstCharacter": false,
    "isLastCharacter": true,
    "tag": "STRONG",
  },
  " ",
  "T",
  "h",
  "i",
  "s",
  " ",
  "i",
  "s",
  " ",
  "s",
  "o",
  "m",
  "e",
  " ",
  "t",
  "e",
  "x",
  "t",
  " ",
  "w",
  "i",
  "t",
  "h",
  " ",
  "H",
  "T",
  "M",
  "L",
  " ",
  "a",
  "t",
  " ",
  "t",
  "h",
  "e",
  " ",
  "b",
  "e",
  "g",
  "i",
  "n",
  "n",
  "i",
  "n",
  "g",
  ".",
]
`;

exports[`Parses HTML tag at end of string. 1`] = `
Array [
  "T",
  "h",
  "i",
  "s",
  " ",
  "i",
  "s",
  " ",
  "s",
  "o",
  "m",
  "e",
  " ",
  "t",
  "e",
  "x",
  "t",
  " ",
  "w",
  "i",
  "t",
  "h",
  " ",
  "H",
  "T",
  "M",
  "L",
  " ",
  "a",
  "t",
  " ",
  "t",
  "h",
  "e",
  " ",
  Object {
    "attributes": Array [],
    "content": "b",
    "isFirstCharacter": true,
    "isLastCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "e",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "g",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "i",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "n",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "n",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "i",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "n",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": "g",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "EM",
  },
  Object {
    "attributes": Array [],
    "content": ".",
    "isFirstCharacter": false,
    "isLastCharacter": true,
    "tag": "EM",
  },
]
`;

exports[`Parses HTML tag with attributes. 1`] = `
Array [
  "T",
  "h",
  "i",
  "s",
  " ",
  "s",
  "t",
  "r",
  "i",
  "n",
  "g",
  " ",
  "h",
  "a",
  "s",
  " ",
  "a",
  "n",
  " ",
  Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "strong-class",
      },
      Object {
        "name": "id",
        "value": "strong-id",
      },
      Object {
        "name": "data-whatever",
        "value": "data-att",
      },
    ],
    "content": "e",
    "isFirstCharacter": true,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "strong-class",
      },
      Object {
        "name": "id",
        "value": "strong-id",
      },
      Object {
        "name": "data-whatever",
        "value": "data-att",
      },
    ],
    "content": "l",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "strong-class",
      },
      Object {
        "name": "id",
        "value": "strong-id",
      },
      Object {
        "name": "data-whatever",
        "value": "data-att",
      },
    ],
    "content": "e",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "strong-class",
      },
      Object {
        "name": "id",
        "value": "strong-id",
      },
      Object {
        "name": "data-whatever",
        "value": "data-att",
      },
    ],
    "content": "m",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "strong-class",
      },
      Object {
        "name": "id",
        "value": "strong-id",
      },
      Object {
        "name": "data-whatever",
        "value": "data-att",
      },
    ],
    "content": "e",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "strong-class",
      },
      Object {
        "name": "id",
        "value": "strong-id",
      },
      Object {
        "name": "data-whatever",
        "value": "data-att",
      },
    ],
    "content": "n",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [
      Object {
        "name": "class",
        "value": "strong-class",
      },
      Object {
        "name": "id",
        "value": "strong-id",
      },
      Object {
        "name": "data-whatever",
        "value": "data-att",
      },
    ],
    "content": "t",
    "isFirstCharacter": false,
    "isLastCharacter": true,
    "tag": "STRONG",
  },
  " ",
  "w",
  "i",
  "t",
  "h",
  " ",
  "a",
  "t",
  "t",
  "r",
  "i",
  "b",
  "u",
  "t",
  "e",
  "s",
  ".",
]
`;

exports[`Parses multiple HTML tags. 1`] = `
Array [
  "H",
  "e",
  "l",
  "l",
  "o",
  ",",
  " ",
  "t",
  "h",
  "i",
  "s",
  " ",
  "i",
  "s",
  " ",
  "s",
  "o",
  "m",
  "e",
  " ",
  Object {
    "attributes": Array [],
    "content": "b",
    "isFirstCharacter": true,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "o",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "l",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "d",
    "isFirstCharacter": false,
    "isLastCharacter": true,
    "tag": "STRONG",
  },
  " ",
  "t",
  "e",
  "x",
  "t",
  ",",
  " ",
  "a",
  "n",
  "d",
  " ",
  "s",
  "o",
  "m",
  "e",
  " ",
  Object {
    "attributes": Array [],
    "content": "i",
    "isFirstCharacter": true,
    "isLastCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "t",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "a",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "l",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "i",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "c",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "i",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "z",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "e",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "I",
  },
  Object {
    "attributes": Array [],
    "content": "d",
    "isFirstCharacter": false,
    "isLastCharacter": true,
    "tag": "I",
  },
  " ",
  "t",
  "e",
  "x",
  "t",
  ".",
]
`;

exports[`Parses normal string correctly. 1`] = `
Array [
  "H",
  "e",
  "l",
  "l",
  "o",
  ",",
  " ",
  "t",
  "h",
  "i",
  "s",
  " ",
  "i",
  "s",
  " ",
  "m",
  "y",
  " ",
  "s",
  "t",
  "r",
  "i",
  "n",
  "g",
  ".",
]
`;

exports[`Parses single HTML tag. 1`] = `
Array [
  "H",
  "e",
  "l",
  "l",
  "o",
  ",",
  " ",
  "t",
  "h",
  "i",
  "s",
  " ",
  "i",
  "s",
  " ",
  "s",
  "o",
  "m",
  "e",
  " ",
  Object {
    "attributes": Array [],
    "content": "b",
    "isFirstCharacter": true,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "o",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "l",
    "isFirstCharacter": false,
    "isLastCharacter": false,
    "tag": "STRONG",
  },
  Object {
    "attributes": Array [],
    "content": "d",
    "isFirstCharacter": false,
    "isLastCharacter": true,
    "tag": "STRONG",
  },
  " ",
  "t",
  "e",
  "x",
  "t",
  ".",
]
`;

exports[`placeholderize() Removes empty tags that live with filled tags. 1`] = `
Object {
  "nodes": Array [
    <em>
      italics!
    </em>,
  ],
  "string": " {%}",
}
`;

exports[`placeholderize() Removes empty tags with text surrounding them. 1`] = `
Object {
  "nodes": Array [],
  "string": "text!  text!",
}
`;

exports[`placeholderize() Removes empty tags. 1`] = `
Object {
  "nodes": Array [],
  "string": "",
}
`;

exports[`placeholderize() Sets placeholder for a self-closing tag with no closing slash. 1`] = `
Object {
  "nodes": Array [
    <br />,
  ],
  "string": "This is a string with a {%} break in it.",
}
`;

exports[`placeholderize() Sets placeholder for a self-closing tag. 1`] = `
Object {
  "nodes": Array [
    <br />,
  ],
  "string": "This is a string with a {%} break in it.",
}
`;

exports[`placeholderize() Sets placeholder for single regular HTML tag. 1`] = `
Object {
  "nodes": Array [
    <strong>
      bold text
    </strong>,
  ],
  "string": "This is a string with some {%}.",
}
`;

exports[`placeholderize() Sets placeholder for strings with both standard tags and self-closing. 1`] = `
Object {
  "nodes": Array [
    <br />,
    <em>
      italized
    </em>,
  ],
  "string": "This is a string with a {%} break in it, as well as some {%} text.",
}
`;

exports[`removePlaceholderRemnants() It does not touch array if last character is not followed by remnants. 1`] = `
Array [
  "x",
  Object {
    "isLastCharacter": false,
  },
  Object {
    "isLastCharacter": false,
  },
  Object {
    "isLastCharacter": true,
  },
  "%",
  "%",
]
`;

exports[`removePlaceholderRemnants() It properly handles arrays completely filled with node objects. 1`] = `
Array [
  Object {
    "isLastCharacter": false,
  },
  Object {
    "isLastCharacter": false,
  },
  Object {
    "isLastCharacter": true,
  },
]
`;

exports[`removePlaceholderRemnants() It properly handles arrays whose node objects are at the very beginning. 1`] = `
Array [
  Object {
    "isLastCharacter": false,
  },
  Object {
    "isLastCharacter": false,
  },
  Object {
    "isLastCharacter": true,
  },
  "%",
  "%",
]
`;

exports[`removePlaceholderRemnants() It removes placeholder remnants following the end of a tag. 1`] = `
Array [
  "x",
  Object {
    "isLastCharacter": false,
  },
  Object {
    "isLastCharacter": false,
  },
  Object {
    "isLastCharacter": true,
  },
]
`;
